File: server/tests/helpers/fixtures.py
import random
import time

import bcrypt

from app.db import query, with_transaction
from app.permissions import flatten_defaults
from app.utils.id import new_id

PASSWORD = "Password123!"

_permissions_seeded = False


def ensure_role_permissions_seeded():
    """Seeds the default role -> permission matrix (idempotent, safe to call from every test)."""
    global _permissions_seeded
    if _permissions_seeded:
        return
    rows = flatten_defaults()

    def _tx(client):
        for r in rows:
            client.query(
                """
                INSERT INTO role_permissions (id, role, resource, action, allowed)
                VALUES (%s, %s, %s, %s, %s)
                ON CONFLICT (role, resource, action) DO UPDATE SET allowed = EXCLUDED.allowed
                """,
                (new_id(), r["role"], r["resource"], r["action"], r["allowed"]),
            )

    with_transaction(_tx)
    _permissions_seeded = True


def _create_user(prefix, name, role, department_id, manager_id=None):
    password_hash = bcrypt.hashpw(PASSWORD.encode("utf-8"), bcrypt.gensalt(12)).decode("utf-8")
    email = f"{prefix}-{name}@test.local"
    user_id = new_id()
    employee_id = new_id()

    query(
        """
        INSERT INTO users (id, employee_code, email, password_hash, role, status, department_id, manager_id)
        VALUES (%s, %s, %s, %s, %s, 'ACTIVE', %s, %s)
        """,
        (user_id, f"{prefix.upper()}-{name.upper()}", email, password_hash, role, department_id, manager_id),
    )
    query(
        """
        INSERT INTO employees (id, user_id, first_name, last_name, department_id, status)
        VALUES (%s, %s, %s, 'Test', %s, 'ACTIVE')
        """,
        (employee_id, user_id, name, department_id),
    )

    return {"id": user_id, "employeeId": employee_id, "email": email, "role": role}


def create_fixture_set(prefix):
    """Creates a fully independent set of fixture users (one per role, plus
    an unrelated manager/employee pair for cross-team scoping tests) under a
    unique prefix, so multiple tests can share the same test database without
    colliding on unique email/employeeCode/department-name constraints."""
    ensure_role_permissions_seeded()

    unique = f"{prefix}-{int(time.time() * 1000)}-{random.randint(100000, 999999)}"
    department_id = new_id()
    query("INSERT INTO departments (id, name) VALUES (%s, %s)", (department_id, f"{unique}-dept"))

    super_admin = _create_user(unique, "superadmin", "SUPER_ADMIN", department_id)
    hr_admin = _create_user(unique, "hradmin", "HR_ADMIN", department_id)
    payroll_admin = _create_user(unique, "payrolladmin", "PAYROLL_ADMIN", department_id)
    manager = _create_user(unique, "manager", "MANAGER", department_id)
    employee = _create_user(unique, "employee", "EMPLOYEE", department_id, manager["id"])
    other_manager = _create_user(unique, "othermanager", "MANAGER", department_id)
    other_employee = _create_user(unique, "otheremployee", "EMPLOYEE", department_id, other_manager["id"])

    return {
        "departmentId": department_id,
        "superAdmin": super_admin,
        "hrAdmin": hr_admin,
        "payrollAdmin": payroll_admin,
        "manager": manager,
        "employee": employee,
        "otherManager": other_manager,
        "otherEmployee": other_employee,
    }

File: server/tests/helpers/fixtures.ts
import bcrypt from 'bcryptjs';
import { prisma } from '../../src/config/prisma';
import { flattenDefaults } from '../../src/permissions';

export const PASSWORD = 'Password123!';

let permissionsSeeded = false;

/** Seeds the default role -> permission matrix (idempotent, safe to call from every test file). */
export async function ensureRolePermissionsSeeded() {
  if (permissionsSeeded) return;
  const rows = flattenDefaults();
  await prisma.$transaction(
    rows.map((r) =>
      prisma.rolePermission.upsert({
        where: { role_resource_action: { role: r.role, resource: r.resource, action: r.action } },
        update: { allowed: r.allowed },
        create: r,
      })
    )
  );
  permissionsSeeded = true;
}

interface FixtureUser {
  id: string;
  employeeId: string;
  email: string;
  role: string;
}

export interface FixtureSet {
  departmentId: string;
  superAdmin: FixtureUser;
  hrAdmin: FixtureUser;
  payrollAdmin: FixtureUser;
  manager: FixtureUser;
  employee: FixtureUser;
  /** A second, unrelated manager+employee pair, useful for cross-team scoping tests. */
  otherManager: FixtureUser;
  otherEmployee: FixtureUser;
}

async function createUser(prefix: string, opts: { name: string; role: string; departmentId: string; managerId?: string }): Promise<FixtureUser> {
  const passwordHash = await bcrypt.hash(PASSWORD, 12);
  const email = `${prefix}-${opts.name}@test.local`;
  const user = await prisma.user.create({
    data: {
      email,
      employeeCode: `${prefix.toUpperCase()}-${opts.name.toUpperCase()}`,
      passwordHash,
      role: opts.role,
      status: 'ACTIVE',
      departmentId: opts.departmentId,
      managerId: opts.managerId,
      employee: {
        create: {
          firstName: opts.name,
          lastName: 'Test',
          departmentId: opts.departmentId,
          status: 'ACTIVE',
        },
      },
    },
    include: { employee: true },
  });
  return { id: user.id, employeeId: user.employee!.id, email: user.email, role: user.role };
}

/**
 * Creates a fully independent set of fixture users (one per role, plus an
 * unrelated manager/employee pair for cross-team scoping tests) under a
 * unique prefix, so multiple test files -- or even multiple calls in the
 * same file -- can share the same test database without colliding on
 * unique email/employeeCode/department-name constraints. The prefix is
 * combined with a random suffix rather than trusted alone, since Vitest
 * may schedule test files across worker processes that don't share the
 * in-memory de-duplication this module would otherwise rely on.
 */
export async function createFixtureSet(prefix: string): Promise<FixtureSet> {
  await ensureRolePermissionsSeeded();

  const unique = `${prefix}-${Date.now()}-${Math.random().toString(36).slice(2, 8)}`;
  const department = await prisma.department.create({ data: { name: `${unique}-dept` } });

  const superAdmin = await createUser(unique, { name: 'superadmin', role: 'SUPER_ADMIN', departmentId: department.id });
  const hrAdmin = await createUser(unique, { name: 'hradmin', role: 'HR_ADMIN', departmentId: department.id });
  const payrollAdmin = await createUser(unique, { name: 'payrolladmin', role: 'PAYROLL_ADMIN', departmentId: department.id });
  const manager = await createUser(unique, { name: 'manager', role: 'MANAGER', departmentId: department.id });
  const employee = await createUser(unique, { name: 'employee', role: 'EMPLOYEE', departmentId: department.id, managerId: manager.id });
  const otherManager = await createUser(unique, { name: 'othermanager', role: 'MANAGER', departmentId: department.id });
  const otherEmployee = await createUser(unique, { name: 'otheremployee', role: 'EMPLOYEE', departmentId: department.id, managerId: otherManager.id });

  return { departmentId: department.id, superAdmin, hrAdmin, payrollAdmin, manager, employee, otherManager, otherEmployee };
}
